### ISSUE TITLE
Create a tutorial to check overlap between irregular shapes using data structure grid

### ISSUE THREAD

**Issue by hepengfe** [asyml/forte#888]
**Is your feature request related to a problem? Please describe.**
One of the main reasons for creating grid is to searching overlapped images efficiently. However, the original code is a lower-level implementation, and it's missing a tutorial of such a usage.

**Describe the solution you'd like**
The full algorithm cannot be complete inside `Grid` alone.
The full algorithm is like below

stage1: build grid cell indices for each image   (use Grid)
```
for img_idx, img in enumerate(images):
      indices = grid.get_overlapped_grid_cell_indices(img) # returns grid cell indices the image is on
      for gc_idx in indices:
            d[gc_idx].append(image_idx)
```

stage2: search for candidate
```
indices = grid.build_index(img)
cand_pool = set()
for gc_idx in indices:
       cand_pool += set(d[gc_idx])
```

stage3: compute iou between the query image and images in cand_pool directly

**Describe alternatives you've considered**

**Additional context**

**Comment by hunterhector** [comment]
Actually, this is not what we discussed, we don't want users to manually compute any `IOU`, they should be able to do something like "a_box.iou(a_circle)". So what we need here is not a tutorial.

1. We want to have a system interanlly to manage stuff like overlap/covering, we don't ask the users to write an algorithm for it. Please make a comparison to the text datapack system, where we can do something like `sentence.get(Token)`. Most of the time the user don't have to interact with the Grid object, just like they don't have to interact with the Span object.
2. Please be careful and precise of the wording, we don't compute "overlapped images", we are computing overalp between ImageAnnotation

@mylibrar

**Comment by mylibrar** [comment]
> Actually, this is not what we discussed, we don't want users to manually compute any `IOU`, they should be able to do something like "a_box.iou(a_circle)". So what we need here is not a tutorial.
> 
> 1. We want to have a system interanlly to manage stuff like overlap/covering, we don't ask the users to write an algorithm for it. Please make a comparison to the text datapack system, where we can do something like `sentence.get(Token)`. Most of the time the user don't have to interact with the Grid object, just like they don't have to interact with the Span object.
> 2. Please be careful and precise of the wording, we don't compute "overlapped images", we are computing overalp between ImageAnnotation
> 
> @mylibrar

Then maybe we should update the title and content of this issue. It should be targeted at implementing some internal algorithms for calculating overlap/covering between `ImageAnnotation`s instead of creating a new tutorial/example.

**Comment by hunterhector** [comment]
> > Actually, this is not what we discussed, we don't want users to manually compute any `IOU`, they should be able to do something like "a_box.iou(a_circle)". So what we need here is not a tutorial.
> > 
> > 1. We want to have a system interanlly to manage stuff like overlap/covering, we don't ask the users to write an algorithm for it. Please make a comparison to the text datapack system, where we can do something like `sentence.get(Token)`. Most of the time the user don't have to interact with the Grid object, just like they don't have to interact with the Span object.
> > 2. Please be careful and precise of the wording, we don't compute "overlapped images", we are computing overalp between ImageAnnotation
> > 
> > @mylibrar
> 
> Then maybe we should update the title and content of this issue. It should be targeted at implementing some internal algorithms for calculating overlap/covering between `ImageAnnotation`s instead of creating a new tutorial/example.

The prioritization of this item can be determined according to other items. This item is a stretch goal for the last quarter. Our goal is to provide computation among arbitrary image annotations, and the defined goal for last quarter is for regular shapes only, like https://github.com/asyml/forte/issues/802

**Comment by hepengfe** [comment]
I think currently our implementation can already compute overlapping between `ImageAnnotation` here https://github.com/asyml/forte/blob/8a61272fe2dd7d3f266f4d66a2f0d3e8f6fc3389/forte/data/ontology/top.py#L936.

For support like `region.get(Box)`, we can build a grid in the DataPack and a dictionary stores image annotation presenting at each grid cell.

**Comment by hunterhector** [comment]
> * 
> * Please be careful and precise of the wording, we don't compute "overlapped images", we are computing overalp between ImageAnnota

Yeah, that means we don't need the tutorial of how to do these for scratch, but it is valuable to show how to use these functions.

For simple `get` with regular shapes, we can do it easily. For irregular shapes, we then need to use `Grid`.

However, most of the supports for irregular shapes are not in our current scope, let's do them later. As for now, if operations involve those, we should throw some NotImplementedError.
